codec/address.go

```diff
@@ -52,30 +59,48 @@ func StringToAddress(s string) (Address, error) {
 
 // String implements fmt.Stringer.
 func (a Address) String() string {
-	return ToHex(a[:])
+	return encodeWithChecksum(a[:])
 }
 
-// MarshalText returns the hex representation of a.
+// MarshalText returns the checksummed hex representation of a.
 func (a Address) MarshalText() ([]byte, error) {
-	result := make([]byte, len(a)*2+2)
-	copy(result, `0x`)
-	hex.Encode(result[2:], a[:])
-	return result, nil
+	return []byte(encodeWithChecksum(a[:])), nil
 }
 
-// UnmarshalText parses a hex-encoded address.
+// UnmarshalText parses a hex-encoded address (with or without a checksum).
 func (a *Address) UnmarshalText(input []byte) error {
-	// Check if the input has the '0x' prefix and skip it
-	if len(input) >= 2 && input[0] == '0' && input[1] == 'x' {
-		input = input[2:]
-	}
-
-	// Decode the hex string
-	decoded, err := hex.DecodeString(string(input))
+	decoded, err := fromChecksum(string(input))
 	if err != nil {
-		return err // Return the error if the hex string is invalid
+		return err
 	}
 
 	copy(a[:], decoded)
 	return nil
 }
+
+// encodeWithChecksum appends a 4 byte checksum and encodes the result in hex format
+func encodeWithChecksum(bytes []byte) string {
+	bytesLen := len(bytes)
+	checked := make([]byte, bytesLen+checksumLen)
+	copy(checked, bytes)
+	copy(checked[AddressLen:], hashing.Checksum(bytes, checksumLen))
+	return "0x" + hex.EncodeToString(checked)
+}
+
+// fromChecksum interprets the hex bytes, validates the 4 byte checksum, and returns the original bytes
+func fromChecksum(s string) ([]byte, error) {
+	if len(s) >= 2 && s[0] == '0' && s[1] == 'x' {
+		s = s[2:]
+	}
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		return nil, err
+	}
+	// Verify checksum
+	originalBytes := decoded[:len(decoded)-checksumLen]
+	checksum := decoded[len(decoded)-checksumLen:]
+	if !bytes.Equal(checksum, hashing.Checksum(originalBytes, checksumLen)) {
+		return nil, ErrBadChecksum
+	}
+	return originalBytes, nil
```

Comment:
Could we handle the case the decoded length is less than the checksum length and add a regression test?

Reply:
Done